core/src/queue.rs

```diff
@@ -137,59 +145,88 @@ impl Queue {
+                let entry = match self.txs.entry(hash) {
+                    Entry::Occupied(mut old_tx) => {
+                        // MST case
+                        old_tx
+                            .get_mut()
+                            .as_mut_v1()
+                            .signatures
+                            .extend(tx.as_v1().signatures.clone());
+                        return Ok(());
+                    }
+                    Entry::Vacant(entry) => entry,
+                };
+                if txs_len >= self.max_txs {
+                    return Err(Failure {
+                        tx,
+                        err: Error::Full,
+                    });
                 }
-                Entry::Vacant(entry) => entry,
+
+                // Insert entry first so that the `tx` popped from `queue` will always have a `(hash, tx)` record in `txs`.
+                entry.insert(tx);
+                let queue_to_push = if signature_check {
+                    &self.queue
+                } else {
+                    &self.waiting_buffer
+                };
+                queue_to_push.push(hash).map_err(|err_hash| {
+                    let (_, err_tx) = self
+                        .txs
+                        .remove(&err_hash)
+                        .expect("Inserted just before match");
+                    Failure {
+                        tx: err_tx,
+                        err: Error::Full,
+                    }
+                })
+            }
+        }
+    }
+
+    /// Pop single transaction from the signature buffer. Record unsigned transaction in `seen`.
+    fn pop_from_signature_buffer(
+        &self,
+        seen: &mut Vec<HashOf<VersionedSignedTransaction>>,
+        wsv: &WorldStateView,
+    ) -> Option<VersionedAcceptedTransaction> {
+        loop {
+            let hash = self.waiting_buffer.pop()?;
+            let entry = match self.txs.entry(hash) {
+                Entry::Occupied(entry) => entry,
+                // FIXME: Reachable under high load. Investigate, see if it's a problem.
+                Entry::Vacant(_) => continue,
             };
 
-            // Reason for such insertion order is to avoid situation
-            // when poped from the `queue` hash does not yet has corresponding (hash, tx) record in `txs`
-            entry.insert(tx);
-            self.queue.push(hash).map_err(|err_hash| {
-                let (_, err_tx) = self
-                    .txs
-                    .remove(&err_hash)
-                    .expect("Inserted just before match");
-                (err_tx, Error::Full)
-            })
+            match self.check_tx(entry.get(), wsv) {
+                Ok(MustUse(signature_check)) => {
+                    // Transactions are not removed from the queue until expired or committed
+                    seen.push(hash);
+                    if signature_check {
+                        return Some(entry.get().clone());
```

Comment:
it might be worth to consider using an `Arc` instead of cloning since what we have is multiple owners (queue and sumeragi) of the same data. This is just a passing thought
 
 this is what I was after when talking about removing transactions from the queue

Reply:
@Erigara says: `i’m not sure that this will work out because we still need cloning becasue block contatin all transactions`

this is a good point, but we have `transaction_cache` containing duplicates of transactions. It might still be worth to consider using `Arc` for that reason. On the other hand, I can hardly believe that `transaction_cache` in sumeragi is making any difference - especially after these changes are merged. We should remove it if possbile

Reply:
Created #3067